docs/proposals/provisioning-proposal.md

```diff
@@ -0,0 +1,112 @@
+# Cluster-API Provisioning Mechanism Consolidation Proposal
+
+## Abstract
+
+Cluster-API requires implementers to implement provisioning mechanisms (bootstrap scripts and logic for running those scripts) themselves. However, the mechanism, including bootstrap scripts, is same or very similar for all cloud providers. As this is a time consuming task it could be very useful to provide a reusable provisioning mechanism in the upstream Cluster-API.
+
+## Proposal
+
+The upstream Cluster-API doesn’t have any bootstrap scripts neither any wrapper around kubeadm for handling cluster operations (initialization and join). This means implementers have to manually write bootstrap scripts used by specific Cluster-API provider and manually handle cluster initialization and node joining. For most cloud providers both bootstrap scripts and steps for cluster initialization/joining are the same, except for provider specifics that could be provided via template variables.
+
+The bootstrap scripts usually do the following tasks:
+
+* Install dependencies required by Kubernetes
+    * Includes dependencies such as socat, ebtables, which are same for all providers.
+* Install container runtime
+    * Steps for installing container runtime such as Docker or containerd are same for all providers.
+* Install Kubernetes packages—kubeadm, kubelet, kubectl, kubernetes-cni
+    * The required packages and steps to install them are same for all providers.
+* Configure kubelet
+    * There are several differences based on provider, however differences could be templated.
+    * kubeadm configuration file can handle some configuration tasks for kubelet.
+* Initialize cluster or join node to a cluster
+    * Should be done using kubeadm configuration files, written by implementer or Cluster-API provider user.
+
+Implementing the provisioning scripts in the upstream Cluster-API ensures:
+
+* Cluster-API implementers can easily get started, as they don’t need to implement complex provisioning logic and scripts for many various setups and operating systems
```

Comment:
@roberthbailey: How many variations do you expect we will maintain in the cluster-api repo?

@xmudrii: I'm not expecting too many variations. For now we should go with basic single-master setup with kubeadm. Regarding Kubernetes versions, depends on how the big difference is and hassle to maintain it. Regarding operating systems, Ubuntu is the most popular one and should be supported, beside it we can choose one more optionally.